experimental/uefi/channel/src/lib.rs

```diff
@@ -32,25 +33,76 @@ pub mod schema {
+type FrameLength = u32;
 const FRAME_LENGTH_SIZE: usize = 4;
 static_assertions::assert_eq_size!([u8; FRAME_LENGTH_SIZE], FrameLength);
 
-/// In request frames this field contains the id of the method_or_status that is invoked.
-/// In response frames this field contains the status code of the method_or_status invocation.
-pub type Method = u32;
+/// In request frames this field contains the id of the method that is invoked.
+/// In response frames this field contains the status code of the method invocation.
+type MethodOrStatus = u32;
 const METHOD_SIZE: usize = 4;
-static_assertions::assert_eq_size!([u8; METHOD_SIZE], Method);
+static_assertions::assert_eq_size!([u8; METHOD_SIZE], MethodOrStatus);
 
-pub const FRAME_HEADER_SIZE: usize = 8;
+// TODO(#2848): Use this space to send a stream identifier that identifies
+// frames as belonging to the same message. Currently the frames of every message
+// must be sent as an uninterrupted seqeuence. A stream identifier would allow us
+// to interleave messages.
+/// Reserved space to maintain 8 byte alignement.
+const SPACER_SIZE: usize = 4;
+
+#[repr(u32)]
+#[derive(Copy, Clone)]
+enum Flag {
+    /// Atomic frame, a message sent as a single frame
+    AtomicFrame = 0,
+    /// Starts a stream of frames that constitute a message
+    StreamStart = 1,
+    /// Continues a stream of frames that constitute a message
+    StreamContinuance = 2,
+    /// Ends a stream of frames that constitute a message, indicating that the recipient may
+    /// process the message.
+    StreamEnd = 3,
+}
+const FLAG_SIZE: usize = 4;
+static_assertions::assert_eq_size!([u8; FLAG_SIZE], Flag);
+impl TryFrom<[u8; FLAG_SIZE]> for Flag {
+    type Error = anyhow::Error;
+
+    fn try_from(value: [u8; FLAG_SIZE]) -> Result<Self, Self::Error> {
+        let flag = match u32::from_le_bytes(value) {
+            0 => Flag::AtomicFrame,
+            1 => Flag::StreamStart,
+            2 => Flag::StreamContinuance,
+            3 => Flag::StreamEnd,
+            _ => anyhow::bail!("invalid flag"),
+        };
+        Ok(flag)
+    }
+}
+impl From<Flag> for [u8; 4] {
+    fn from(flag: Flag) -> Self {
+        (flag as u32).to_le_bytes()
+    }
+}
+
+const FRAME_HEADER_SIZE: usize = 16;
 static_assertions::assert_eq_size!(
     [u8; FRAME_HEADER_SIZE],
-    [u8; FRAME_LENGTH_SIZE + METHOD_SIZE]
+    [u8; FRAME_LENGTH_SIZE + METHOD_SIZE + SPACER_SIZE + FLAG_SIZE]
 );
+// Ensure that the frame header is 8 Byte aligned.
+static_assertions::const_assert!(FRAME_HEADER_SIZE % 8 == 0);
 
-pub struct Frame {
-    pub method_or_status: Method,
-    pub body: Vec<u8>,
+const MAX_FRAME_SIZE: usize = 64000;
+const MAX_FRAME_BODY_SIZE: usize = MAX_FRAME_SIZE - FRAME_HEADER_SIZE;
+
+/// A [`Frame`] is a small unit data that is sent over the communication
+/// channel. Usually several [`Frame`]'s are used to send a [`Message`].
+#[derive(Clone)]
+struct Frame {
+    method_or_status: MethodOrStatus,
+    flag: Flag,
+    body: Vec<u8>,
```

Comment:
Perhaps we should distinguish Request and Response frames at some point, instead of overloading the semantics of method_or_status and flag fields?

Reply:
Maybe, imo it should be fine for now. Having a single struct makes it a lot more straightforward to do the logic for message encoding and message fragmentation, both of which is shared. Flags should also be the same across requests and responses